GIT-6: Clone non-public repositories

Haxsen opened the pull request:
[Issue Link](https://jira.xwiki.org/projects/GIT/issues/GIT-6)

Glad to correct my mistakes.
Thanks for your reviews and valuable time :)

* Before:
![PrivateGitBefore](https://user-images.githubusercontent.com/33906649/78502269-c189c080-7779-11ea-871f-5bb79f83ed87.png)

* After:
![PrivateGitAfter](https://user-images.githubusercontent.com/33906649/78502270-c5b5de00-7779-11ea-8e5c-a5522a36e774.png)

Haxsen committed:
GIT-6: Clone non-public repositories

Add a private getGitRepository method to combine the code for getRepository and getPrivateRepository.
Create cloneCommand to minimize calling.
[Misc] Add comments and some corrections.

Haxsen committed:
GIT-6: Clone non-public repositories

Add `getPrivateRepository` method for private repositories.

Haxsen committed:
GIT-6: Clone non-public repositories

Add getPrivateRepository method for cloning private repositories which need credentials.
Prevent breaking backward compatibility.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -48,6 +48,24 @@
+     * @param username the username of Git user
+     * @param token the Personal access token generated by the user
+     * @return the cloned Repository instance
+     * @since 12.3RC1
```
This is not correct. I doubt that the version of the Git extension is at 12.3RC1 :)

Haxsen replied:
Yeah I was wondering about this if it's either the version of the extension or the xwiki.
Updated.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -48,6 +48,24 @@
+    default Repository getPrivateRepository(String repositoryURI, String localDirectoryName,
+                                            String username, String token)
+    {
+        throw new UnsupportedOperationException();
```
Cool, you found it :)

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -48,6 +48,24 @@
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of Git user
+     * @param token the Personal access token generated by the user
```
Maybe mention how to get that token. Hmmm what token is this about?

Haxsen replied:
Updated.

vmassol commented on src/main/java/org/xwiki/git/internal/DefaultGitManager.java:
```diff
@@ -67,32 +69,42 @@
-    @Override
-    public Repository getRepository(String repositoryURI, String localDirectoryName)
+    private Repository getGitRepository(String repositoryURI, String localDirectoryName,
+                                        String username, String token)
```
For indentations and code style please see https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/

Haxsen replied:
Updated. Please confirm. (will update the others then)

vmassol commented on src/main/java/org/xwiki/git/internal/DefaultGitManager.java:
```diff
@@ -67,32 +69,42 @@
         this.logger.debug("Local Git repository is at [{}]", gitDirectory);
-        FileRepositoryBuilder builder = new FileRepositoryBuilder();
 
+        // Initialize builder for the Git environment.
```
I think this comment is not really useful, is it? :)

vmassol commented:
@Haxsen the PR is good. I have only small comments that I have put. Cool.

One thing though, the test has not been updated with a new test to verify that it works.

Haxsen committed:
GIT-6: Clone non-public repositories

Fix @since version and mention where the token can be generated.

Haxsen commented:
@vmassol Yep I thought it might not have tests since the repository is private (and need that user's credentials).
I will update you with the tests.

vmassol commented:
The test creates the git repo. I still don't know what a personal access token is for git. We're talking about git, not github, right?

Haxsen committed:
GIT-6: Clone non-public repositories

Update the codestyle (for the new method) and javadoc for @param token.

Haxsen commented:
@vmassol  Yes, not just for github. Personal access token is required to authenticate with the user ID. It also works with Two-factor authentication. See [Gitlab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) and [Github](https://developer.github.com/v3/auth/#working-with-two-factor-authentication) pages about it.
Forgot that there's also others like GitLab, my mistake.
Updated the javadoc.
Let me know if there are more mistakes.

vmassol commented:
> Yes, not just for github. Personal access token is required to authenticate with the user ID. It also works with Two-factor authentication. See Gitlab and Github pages about it.

I don't understand it. This is a git module not a gitlab or github module. If I create a git repo locally, it should work. I admit that I don't know how git permissions work.

EDIT: I guess the concept of private git repo only exists for the HTTP protocol and there's a generic spec for securing a remote git repo. Would you have a link?

vmassol commented:
Just checked UsernamePasswordCredentialsProvider at https://github.com/eclipse/jgit/blob/master/org.eclipse.jgit/src/org/eclipse/jgit/transport/UsernamePasswordCredentialsProvider.java and it refers to username and passwords, not token. So I have the feeling that the API you added is not correct and that the javadoc you added is even more wrong :)

Meaning that it'll work with any remote server that protects a git repo, whether it uses tokens, simple passwords, etc.

Haxsen commented:
@vmassol 
On the [Github page](https://developer.github.com/v3/auth/#working-with-two-factor-authentication) I linked, it says:
"Deprecation Notice: GitHub will discontinue password authentication to the API."
If we use password authentication, it will deprecated.
So what do you suggest? Will follow your guide.

vmassol commented:
@Haxsen I'll repeat it once more: we are NOT doing an API for GitHub or GitLab. We're writing a generic Git API. I just want you to understand what you're doing... you need to understand why your javadoc is just wrong. That was my original question, remember?  It's definitely not about what I want or don't want. I'm not the one writing the git spec!

Haxsen commented:
@vmassol Ok, give me some time I will research more into this.
Will update you. Thanks.

Haxsen committed:
GIT-6: Clone non-public repositories

Change param 'token' to 'accessCode' which can be password/oauth/token provided by user.
Updated javadoc.

Haxsen committed:
GIT-6: Clone non-public repositories

Change param 'token' to 'accessCode'.
Fix indentation.

Haxsen committed:
GIT-6: Clone non-public repositories

Change param 'token' to 'accessCode' and update the javadoc.
Fix indentation.

Haxsen committed:
GIT-6: Clone non-public repositories

Change 'token' to 'accessCode'.

Haxsen commented:
@vmassol 
Ah I understand it now, I feel dumb for using token as param.
Now I have used `accessCode` which is more generic and can be password/oauth/token 😄 
Updated.
Let me know if I did it right or it has more problems.
I apologize for the frustration :/

Haxsen committed:
GIT-6: Clone non-public repositories

Fix indentation.

Haxsen committed:
GIT-6: Clone non-public repositories

Fix @since in javadoc.

Haxsen committed:
GIT-6: Clone non-public repositories

Add Unstable notation.

Haxsen committed:
GIT-6: Clone non-public repositories

Add Unstable notation.

vmassol commented on src/main/java/org/xwiki/git/GitManager.java:
```diff
@@ -48,6 +49,25 @@
+     */
+    @Unstable
+    default Repository getPrivateRepository(String repositoryURI, String localDirectoryName, String username,
+            String accessCode)
```
just 4 spaces indentation. Could you fix all places please? Thx

Haxsen replied:
Sure.

Haxsen replied:
Done @vmassol . Can you confirm if I did right?

vmassol commented:
Looks good (except the indentation)! :)We're getting there

Haxsen committed:
GIT-6: Clone non-public repositories

Fix indentation in GitManager.

Haxsen committed:
GIT-6: Clone non-public repositories

Fix indentation in DefaultGitManager.

Haxsen committed:
GIT-6: Clone non-public repositories

Fix indentation in GitScriptService.

Haxsen commented:
Fixing indentation, also: I have XWiki IntelliJ codestyle set up from the Java code style page. I press enter and it indents 8 spaces instead of 4.

vmassol commented:
> Fixing indentation, also: I have XWiki IntelliJ codestyle set up from the Java code style page. I press enter and it indents 8 spaces instead of 4.

Weird. Maybe that intellij codestyle is wrong? Could you check? I looked at https://raw.githubusercontent.com/xwiki/xwiki-commons/master/xwiki-commons-tools/xwiki-commons-tool-verification-resources/src/main/resources/codestyle-idea.xml and I only see the number 4. No 8.

Haxsen commented:
It's 4 on my end as well. I have restarted my IntelliJ several times. (Let me try Invalidate Caches and Restart).

vmassol commented:
Applied, thanks!

Haxsen commented:
You're welcome and thanks!

vmassol commented:
Actually it's not good yet... You forgot to run the build with the quality profile. See https://ci.xwiki.org/view/Contrib/job/XWiki%20Contrib/job/api-git/job/master/36/console . You dropped the quality of the module, hence the build failure.

> [WARNING] Rule violated for bundle api-git: instructions covered ratio is 0.81, but expected minimum is 0.87

Haxsen commented:
@vmassol 
I see, hm. Let me learn about this problem. I would try to reproduce this issue then fix. I think it might be due to tests, but let's see.
Thanks for letting me know.

vmassol commented:
It is due to the tests covering less code than before since you added code but not tests :)

Haxsen commented:
@vmassol Thanks for the hint. I am working on test now.

Haxsen commented:
Hi @vmassol , I fixed indentation manually by putting 4 in codestyle settings (in IDEA) as it picked 8 (maybe IDEA bug).
I made tests and built on windows but windows didn't throw coverage errors. So I had to reproduce the issue in order to fix. I found that windows has many problems than linux in terms of maven building (and problems with java files, code styles etc. as well). So I had to install linux and test on it.
Just to mention, openJDK latest version fails to build (java crashes) so I used openJDK 8 instead.
PR sent for tests.
Thanks.
